### PR TITLE
Add support for 'new.target' meta-property

### LINKED ISSUE #2551: Support 'new.target' meta object
There isn't a lot in the current ES6 spec draft, but it seems to indicate the constructor object that is in the process of being `new`'d at runtime.

### PR DESCRIPTION
**Pull request by rbuckton**
This adds support for the `new.target` meta-property introduced in ES6. This also adds best-effort down-level emit for ES5/3.

# NewTarget meta-property

The `new.target` meta-property is new syntax introduced in ES6. When you create an instance of a constructor via `new`, the value of `new.target` is set to be a reference to the constructor function initially used to allocate the instance. If a function is called rather than constructed via `new`, `new.target` is set to `undefined`.  In addition, `new.target` is lexically scoped to the body of the function. As such, it can be captured inside an arrow function.

While it is legal to use `new.target` in methods and accessors in ES6, those members cannot be called via `new` as they are not constructors. As such, we have chosen to make it an error to reference `new.target` in the body of a method or accessor.

# ES5 emit

`new.target` is a pure ES6 syntactic construct that can only be emitted down-level on a best-effort basis. While our emit for `new.target` is sufficient for ES5 code that interacts with other ES5 code, it does not fully emulate `new.target` in an ES6 host environment. What this means is that our `new.target` emit is not compatible with `Reflect.construct` as we have no way to intercept this API. This emit also is intolerant of explicit changes to the `constructor` property of the constructor's prototype.

That said, the reason we are introducing best-effort down-level emit for `new.target` are for several specific use cases pairing ES5 emit for classes with `Object.setPrototypeOf` or `__proto__`. One such use case is inheriting from `Error` in NodeJS v4 and higher:

```ts
// ts
class CustomError extends Error {
  constructor(message?: string) {
    super(message); // 'Error' breaks prototype chain here
    (<any>this).__proto__ = new.target.prototype; // restore prototype chain
  }
}

// js
var CustomError = (function (_super) {
  __extends(CustomError, _super);
  function CustomError() {
    var _newTarget = this.constructor;
    var _this = _super.apply(this, arguments);  // 'Error' breaks prototype chain here
    _this.__proto__ = _newTarget.prototype; // restore prototype chain
    return _this;
  }
  return CustomError;
})(Error);
```

Also, while we can use `this.constructor` to emulate `new.target` for classes, we must be more defensive for `new.target` in a function declaration or function expression. The reason for this is that calling an imported function in a CommonJS, AMD, or System module results in the `this` argument of the function being bound to the module object (i.e. `func()` becomes `package_1.func()`). Also, function expressions are often assigned to a property of an object which could in turn become the `this` if the function is called as a member.

To defend against this case, the down-level emit for `new.target` in functions is the following:

```ts
// ts
function f() {
  if (new.target) { /* called via 'new' */ }
}

// js
function f() {
  var _newTarget = this && this instanceof f ? this.constructor : void 0;
  if (_newTarget) { /* called via 'new' */ }
}
```

Fixes #2551 

Related: 
 * https://tc39.github.io/ecma262/#prod-NewTarget - Specification for NewTarget 
 * https://github.com/Microsoft/TypeScript-wiki/pull/100 - Documented breaking change for subclassing built-ins
 * #12660 - Wrong constructor of Error subclass from 2.1.1
 * #12669 - Bug or breaking change in super constructors?

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I'll have to dive in, but can you explain the es2015 transform changes?

File: src/compiler/transformers/es2015.ts [hunk 1]
```diff
@@ -162,6 +162,82 @@ namespace ts {
...
+        IterationStatementBlock = 1 << 9,           // Enclosing Block is enclosed in an IterationStatement
+        ForStatement = 1 << 10,                     // Enclosing block-scoped container is a ForStatement
+        ForInOrForOfStatement = 1 << 11,            // Enclosing block-scoped container is a ForInStatement or ForOfStatement
+        ConstructorWithCapturedSuper = 1 << 12,     // Enclosed in a constructor that captures 'this' for use with 'super'
+        ComputedPropertyName = 1 << 13,             // Enclosed in a computed property name
+        AncestorFactsMask = (ComputedPropertyName << 1) - 1,
+
+        // Subtree facts
+        NewTarget = 1 << 14,                        // Contains a 'new.target' meta-property
+        NewTargetInComputedPropertyName = 1 << 15,  // Contains a 'new.target' meta-property in a computed property name.
```
Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: When we compute flags for a member, such as a method or accessor, it both has its own lexical scope (for the parameters and body) as well as shares the lexical scope of its container (due to the computed property name). Without this flag, encountering a NewTarget in a computed property name would incorrectly indicate that NewTarget was set in the body of the method, not in the container.

**Comment by rbuckton** [comment]
Regarding the changes to the transformer, I am trying to clean up the es2015 transformer and remove some of the variables we use to track state as we descend the tree, this should reduce the overall memory footprint due to values saved on the stack.

File: src/compiler/transformers/es2015.ts [hunk 2]
```diff
@@ -1357,6 +1401,73 @@ namespace ts {
...
+                            createPropertyAccess(
+                                setEmitFlags(createThis(), EmitFlags.NoSubstitution),
+                                "constructor"
+                            ),
+                            createVoidZero()
+                        );
+                        break;
+                }
+
+                const captureNewTargetStatement = createVariableStatement(
```
Comment (DanielRosenwasser) [inline src/compiler/transformers/es2015.ts]: `newTarget` could be `undefined` if you're somehow not handling a case - consider adding a

```ts
Debug.assert(!!newTarget, "'newTarget' should be defined for " + (ts.SyntaxKind as any)[node.kind])
```

or something similar.

File: src/compiler/transformers/es2015.ts [hunk 3]
```diff
@@ -468,76 +477,105 @@ namespace ts {
...
 
         function visitThisKeyword(node: Node): Node {
-            Debug.assert(convertedLoopState !== undefined);
-            if (enclosingFunction && enclosingFunction.kind === SyntaxKind.ArrowFunction) {
-                // if the enclosing function is an ArrowFunction is then we use the captured 'this' keyword.
-                convertedLoopState.containsLexicalThis = true;
-                return node;
+            if (convertedLoopState) {
+                if (hierarchyFacts & HierarchyFacts.ArrowFunction) {
+                    // if the enclosing function is an ArrowFunction is then we use the captured 'this' keyword.
```
Comment (DanielRosenwasser) [inline src/compiler/transformers/es2015.ts]: "is then" -> ", then"

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -900,6 +900,17 @@ namespace ts {
         return false;
     }
 
+    export function unwrapInnermostStatmentOfLabel(node: LabeledStatement, beforeUnwrapLabelCallback?: (node: LabeledStatement) => void) {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: You only use this function once - do we anticipate using it again?

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -3848,6 +3874,7 @@ namespace ts {
             || kind === SyntaxKind.TrueKeyword
             || kind === SyntaxKind.SuperKeyword
             || kind === SyntaxKind.NonNullExpression
+            || kind === SyntaxKind.MetaProperty
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Do we want to just make this a `switch`? I think it's asking for it at this point.

**Comment by DanielRosenwasser** [comment]
I get the crux of the changes - I'm not an expert at the loop conversion code, so @vladima may be a good person to check in with.

File: src/compiler/utilities.ts (hunk 4 above)
Comment (rbuckton) [inline src/compiler/utilities.ts]: Yes, I also am using this in the asyncGenerators branch

**Pull request merged**